Add CUDA Vital

bitfort opened the pull request:
Summary: Add CUDA.used vital sign which is true only if CUDA was "used" which technically means the context was created.

Test Plan: buck test mode/dbg caffe2/test:torch -- --regex basic_vitals

Differential Revision: D28357615

xuzhao9 commented on aten/src/ATen/core/Vitals.cpp:
```diff
@@ -34,11 +38,31 @@ bool torchVitalEnabled() {
+
+  std::stringstream buf;
+  for (auto x : name_map_) {
+    buf << x.second.toString();
```
Why we are only returning the "second" values here, but not "first"? In the test below, it looks like both the name and the value should be returned, say 'CUDA.used\t\t true'.

xuzhao9 commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
```diff
@@ -52,6 +53,9 @@ std::unique_ptr<THCState, void (*)(THCState*)> CUDAHooks::initCUDA() const {
   THCState* thc_state = THCState_alloc();
 
+  // TODO Set Vital to True
+  at::vitals::VitalsAPI.setVital("CUDA", "used", "true", /* force = */ true);
```
May I ask what is the field of "force" doing here?

xuzhao9 reviewed (commented):
I am not familiar with the Vitals project, so just adding a few questions here.

xuzhao9 reviewed (approved):
Looking good, having two comments inlined

ngimel commented on aten/src/ATen/core/Vitals.cpp:
```diff
@@ -51,9 +75,19 @@ bool APIVitals::setVital(
 
-  vital->create(attr_name) << value;
+  if (force) {
+    vital->create(attr_name, /* force = */ true).force_write(value);
```
here conditional also seems like unnecessary complication - if you had `write(value, force)` you could have avoided conditional

ngimel commented on aten/src/ATen/core/Vitals.h:
```diff
@@ -25,6 +26,20 @@ struct TORCH_API TorchVitalAttr {
   }
+
+  template <typename T>
+  void write(const T& t) {
```
Why are there 2 functions instead of `write(const T& t, bool force)`?

ngimel commented on aten/src/ATen/core/Vitals.h:
```diff
@@ -35,11 +50,19 @@ struct TORCH_API TorchVital {
+  TorchVitalAttr& create(const std::string& attr, bool force);
 
-  ~TorchVital() {
+  std::string toString() {
```
Why toString function instead of redefining `<<` so that you can directly call std::cout << this ?

Commit:
Add CUDA Vital (#58059)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/58059

Add CUDA.used vital sign which is true only if CUDA was "used" which technically means the context was created.

Also adds the following features:
- Force vitals to be written even if vitals are disabled, to enable testing when the env variable is not set from the start of execution
- Add a read_vitals call for python to read existing vital signs.

Test Plan: buck test mode/dbg caffe2/test:torch -- --regex basic_vitals

Reviewed By: xuzhao9

Differential Revision: D28357615

fbshipit-source-id: 78a8755801b11ccc7e33a9eac669ebb7a83ddb87

xuzhao9 reviewed (approved):
LGTM!
